Content.Server/Body/Behavior/LiverBehavior.cs

```diff
@@ -90,6 +95,10 @@ public override void Update(float frameTime)
                     continue;
                 }
 
+                // How much reagent is available to metabolise?
+                // This needs to be passed to other functions that have metabolism rate information, such that they don't "overmetabolise" a reagant.
+                var availableReagent = bloodstream.Solution.Solution.GetReagentQuantity(reagent.ReagentId);
+
```

Comment:
rather than relying on the metabolism behavior to check that it doesnt 'overmetabolize', shouldnt metabolizable/liver just check how much is returned and clamp it appropriately? i still think that available reagent amount should be passed in for stuff like chemicals that scale their power based off of reagent quantity, but i dont think that this is something that the behaviors should have to worry about handling

Reply:
The issue is that metabolizable.Metabolize() determines how much should be metabolized using the metabolization rate, AND applies the effect, the liver just removes the reagent from the solution. If you want to clamp the ammount in the liver, you either need to also apply the effect in the liver, or already know the metabolization rates for each reagent beforehand. Without reworking how liver/metabolization works I think this is the easiest way to go about it. And as you said, for things that scale with reagent or have overdosing effects, passing the quantity would be needed.

Reply:
Just to make it clearer: the ''overmetabolizing" issue was not that the liver was removing more reagent than there was available -- it just gets set to zero in that case. The issue was thet metabolizable.Metabolize() applies the effect of the metabolism, and  that effect might have been stronger than should be allowed if only a small amount of reagent is left,

Reply:
ah that makes sense, in that case you can keep the checks for the metabolisms that actually apply effects based on the amount left, though I still dont like the code duplication

Reply:
So I believe the correct approach to avoid code duplication here would be to just have this check in DefaultMetabolizable, and have the other existing Metabolism Classes inherit the checks & tickTime*reactionRate calculation from that. That way that logic only appears in one place, and any metabolism using a different rate calculation would not be a 'DefaultMetaboizable'. I still know very little about the design (prototypes, ECS, etc). Would that be the right way of doing this?

Also, currently each class has its own variables that are filled in from the yaml file. I have no idea how that works. If they were to inherit from DefaultMetabolizable, would they still automatically fill in the dataFields defined in DefaultMetabolizable, or would I still habe to define a [DataField("rate")] in each inheriting class?

Reply:
I have pushed two commits that address this, but I am not sure if the changes are in line with the coding practices that should be used, someone should check that.